### ISSUE TITLE
Request: Clarification on daemon vs worker / non-systemd usage

### ISSUE THREAD

**Issue by AndydeCleyre** [federico-terzi/espanso#302]
I am using s6 to manage espanso. Based on what I saw in the `.service` file, I had my process tree like this:

```
s6-svscan /home/andy/svcs
  |-s6-supervise espanso
  |   `-espanso daemon
  |       `-espanso worker
```

But then I realized that whenever I sent the `espanso daemon` process a `SIGTERM`, and maybe sometimes when using `espanso edit`, the worker process would get disowned and then duplicated.

1. Should `daemon` send a `SIGTERM` to `worker` if it receives that signal itself?

The `worker` subcommand isn't in the `--help` output, and the current _description_ for `daemon` seems like what I want, but now I think `worker` is probably what I want to use directly under a supervisor, and that the `daemon` subcommand is probably trying to act as a supervisor itself, ensuring the `worker` is always up.

2. Is that accurate?

So I've cut `espanso daemon` out of the process tree.

3. If not using systemd for espanso, I will want to avoid these, right?:
- `espanso edit`'s restart feature
- subcommands:
  - `cmd`
  - `log`
  - `register`
  - `restart`
  - `start`
  - `status`
  - `stop`
  - `unregister`

Thanks again for the great project!

**Comment by federico-terzi** [comment]
Hey,

> I am using s6 to manage espanso. Based on what I saw in the `.service` file, I had my process tree like this:
> 
> ```
> s6-svscan /home/andy/svcs
>   |-s6-supervise espanso
>   |   `-espanso daemon
>   |       `-espanso worker
> ```
> 
> But then I realized that whenever I sent the `espanso daemon` process a `SIGTERM`, and maybe sometimes when using `espanso edit`, the worker process would get disowned and then duplicated.
> 
> 1. Should `daemon` send a `SIGTERM` to `worker` if it receives that signal itself?

This is definitely a good point. I'll have to think about it carefully, but this is most probably a missing piece of the implementation. Thanks for pointing that out!

> The `worker` subcommand isn't in the `--help` output, and the current _description_ for `daemon` seems like what I want, but now I think `worker` is probably what I want to use directly under a supervisor, and that the `daemon` subcommand is probably trying to act as a supervisor itself, ensuring the `worker` is always up.
> 
> 1. Is that accurate?

Not quite. The whole daemon+worker mechanism is used to implement the "auto-restart" feature introduced in 0.6.0. Basically the daemon spawns the worker and looks for changes in the config directory. If there are some changes, the worker is respawned. But if the worker exits, the daemon doesn't attempt to respawn it.

> So I've cut `espanso daemon` out of the process tree.

The current implementation assumes the daemon is running as well, so some things might not work as expected without it.

> 1. If not using systemd for espanso, I will want to avoid these, right?:
> 
> * `espanso edit`'s restart feature
> * subcommands:
>   
>   * `cmd`
>   * `log`
>   * `register`
>   * `restart`
>   * `start`
>   * `status`
>   * `stop`
>   * `unregister`
> 
> Thanks again for the great project!

Mostly yes, even though the `log` and `status` commands should work in any scenario, even when not using Systemd.

When adding the "auto-restart" feature I knew there could be some problems (even though the feature itself is so useful in most cases that all the trouble was worth it). As you suggested, we might need to juggle a bit with SIGNAL propagations to obtain a uniform mechanism.

To recap, the expected behaviour would be:
1. If the daemon process receives a SIGTERM, the signal should be propagated to the worker as well
2. Also the other way around (worker -> daemon).

Do you think this would be reasonable?

Cheers :)

**Comment by AndydeCleyre** [comment]
Thanks for the explanations!

I'd think:

`SIGTERM` -> `daemon` should send `SIGTERM` -> `worker`

Maybe same for `SIGKILL`, maybe not. But I really can't say about the other direction. To me, the auto-restart feature seems to be in the domain of process supervision, for which I'd use a "true" supervisor. Whether it intends to or not, that `daemon` process is in the _position_ of a supervisor, and without being a fully fledged one, constitutes a less predictable link in the supervision chain.

There might be some fancy systemd options to use within the `.service` file to restart the `worker` if it's running and the configs change.

For my own stubborn ways, to achieve that behavior, I'd probably have:

- a supervised `worker`
- another supervised process that watches the configs until they change, at which point it sends a `SIGTERM` to the supervised `worker` and returns (letting the supervisor then restart them both)

I understand if having the pseudo-supervisor process is the best option for general distribution, but I'd like to work around that if I can; to that end:

> The current implementation assumes the daemon is running as well, so some things might not work as expected without it.

Anything particular to look out for? Aside from the "avoid" list above, I can guess I'll lose notifications.

P.S. A quick search for doing this with systemd shows [a similar multi-service solution](https://superuser.com/questions/1171751/restart-systemd-service-automatically-whenever-a-directory-changes-any-file-ins), but that answer is old and there may be a newer way by now.

**Comment by AndydeCleyre** [comment]
In case anyone looking to do something similar finds this,
here's an example setup with s6 and inotify-tools.

`/home/andy/svcs/espanso/run`:
```bash
#!/bin/execlineb -P
fdmove -c 2 1  # Send stderr to stdout
espanso worker
```

`/home/andy/svcs/espanso/log/run`
```bash
#!/bin/execlineb -P
s6-log                          # Receive process output
  T                             # Start each line with an ISO 8601 timestamp
  s4194304                      # Archive log when it gets this big (bytes)
  S41943040                     # Purge oldest archived logs when the archive gets this big (bytes)
  /home/andy/svcs-logs/espanso  # Store logs here
```

`/home/andy/svcs/espanso-cfg-watch/run`:
```bash
#!/bin/execlineb -P
fdmove -c 2 1                      # Send stderr to stdout
foreground {
    inotifywait                    # Wait for one of these events
        -e modify
        -e attrib
        -e move
        -e move_self
        -e create
        -e delete
        -e delete_self
        -e unmount
        -r /home/andy/.config/espanso
}
s6-svc -t /home/andy/svcs/espanso  # Send TERM to the espanso worker
```

`/home/andy/svcs/espanso-cfg-watch/log/run`:
```bash
#!/bin/execlineb -P
s6-log                                    # Receive process output
  T                                       # Start each line with an ISO 8601 timestamp
  s4194304                                # Archive log when it gets this big (bytes)
  S41943040                               # Purge oldest archived logs when the archive gets this big (bytes)
  /home/andy/svcs-logs/espanso-cfg-watch  # Store logs here
```

**Comment by federico-terzi** [comment]
Hey @AndydeCleyre 

Thank you for the detailed explanation, I'm sure many users will find it valuable!
As you mentioned, the daemon+worker implementation was necessary to make it cross-platform, as espanso can't depend on the specific functionality of a daemon manager.

That said, I think your solution will work mostly good! You might get a different log, and the list of commands you mentioned above will not work, but for the main features, it should work fine.

> SIGTERM -> daemon should send SIGTERM -> worker

I'll make sure to add this feature in one of the future releases
PS: after we implement this signal propagation, there shouldn't be any reason to avoid using the daemon for you right?

Let me know if you have other doubts

Cheers :)

**Comment by AndydeCleyre** [comment]
Thanks very much!

I will probably continue to use `espanso worker` directly, as the benefits of using `espanso daemon` are for the most part things the supervisor will offer anyway, and I'll need to be using a supervisor in either case. And I like using a consistent interface for managing all similarly supervised processes, rather than special-casing espanso. Also, the chain will still be more reliable -- less risk of getting a loose extra worker running around.

**Comment by AndydeCleyre** [comment]
FWIW there are some cases where the worker dies, maybe for example when an unusual window type comes up; I think I just triggered this with the login password entry window for [rbw](https://github.com/doy/rbw), and its log shows:

```
2020-06-04 14:05:35.486462121  X Error of failed request:  BadWindow (invalid Window parameter)
2020-06-04 14:05:35.486613348    Major opcode of failed request:  20 (X_GetProperty)
2020-06-04 14:05:35.486618125    Resource id in failed request:  0x2a00003
2020-06-04 14:05:35.486621977    Serial number of failed request:  23
2020-06-04 14:05:35.486640823    Current serial number in output stream:  23
```

So the fact that

> . . . if the worker exits, the daemon doesn't attempt to respawn it

suggests that if I had my supervisor process directly parenting the daemon, I'd have to manually trigger a restart of that daemon; while directly parenting the worker means that it's immediately started again in such a case.

**Comment by federico-terzi** [comment]
Sounds definitely reasonable! And thanks for pointing the other error out :)
